### PR TITLE
Add get all products endpoint

### PR DESCRIPTION
**Pull request by Darvand**
## Pull request type

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Missing products controller, services, mapper, dtos and exceptions

Issue Link: https://trello.com/c/6C7EU7fj/12-as-a-user-i-should-be-able-to-see-a-list-of-products

## What is the new behavior?

- Add controller, services, mapper, dto, and exceptions
- Update migrations
- Add tests for new behavior

## Other information

- Pagination and Swagger documentation will be added in the next PR
N/A

## Preview

image

### CODE REVIEW

File: src/products/dtos/product.dto.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+export class ProductDTO {
```
Comment (slasia) [inline src/products/dtos/product.dto.ts]: You can use `class-validator` library  to validate this fields. for example ` @IsNotEmpty() @IsString() @MaxLength(value)` etc...

File: src/products/entities/product.entity.ts [hunk 2]
```diff
@@ -4,18 +4,33 @@ import { Base } from '@shared/entities/base.entity';
 
 @Entity('products')
 export class ProductEntity extends Base {
-  @Column({ type: 'varchar', unique: true, length: 255, nullable: false })
```
Comment (slasia) [inline src/products/entities/product.entity.ts:7]: Why did you remove the length property?

File: src/products/mappers/products.mapper.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { ProductDTO } from '@products/dtos/product.dto';
+import { ProductEntity } from '@products/entities/product.entity';
+
+export class ProductsMapper {
+  static entityToDTO(entity: ProductEntity) {
+    return new ProductDTO(
+      entity.name,
```
Comment (slasia) [inline src/products/mappers/products.mapper.ts]: You can use destructuring assignment like: 


`const {name, description, price,stock,image} = entity`

File: package.json [hunk 4]
```diff
@@ -52,6 +54,7 @@
     "husky": "^8.0.0",
     "jest": "28.0.3",
     "prettier": "^2.3.2",
+    "faker": "^5.5.3",
```
Comment (EMazzaglia) [inline package.json]: I strongly recommend to start using `@faker-js/faker` instead since it's now the official repository. 
Also you dont need the types :)

File: src/products/dtos/product.dto.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  readonly name: string;
```
Comment (EMazzaglia) [inline src/products/dtos/product.dto.ts]: Any of these properties have `class-validator` validations, hence, we can create products without any restriction. Remember to add those in each DTO that you are creating :)

File: src/products/entities/product.entity.ts [hunk 6]
```diff
@@ -4,18 +4,33 @@ import { Base } from '@shared/entities/base.entity';
...
-  stock: number;
+  readonly stock: number;
 
-  @Column({ type: 'varchar', nullable: true, length: 500 })
-  image: string;
+  @Column({ type: 'varchar', nullable: true })
+  readonly image: string;
+
+  constructor(
+    name: ProductEntity['name'],
```
Comment (EMazzaglia) [inline src/products/entities/product.entity.ts]: Can you explain to me what's this syntax doing?

File: src/products/exceptions/get-all-products.exception.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import { HttpException, HttpStatus, Logger } from '@nestjs/common';
```
Comment (EMazzaglia) [inline src/products/exceptions/get-all-products.exception.ts:1]: I suggest to chose a naming convention for Files, and stick to it in the whole project. Otherwise you are having camelCase in some filenames, and hypen-case in others.

File: src/products/mappers/products.mapper.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { ProductDTO } from '@products/dtos/product.dto';
+import { ProductEntity } from '@products/entities/product.entity';
+
+export class ProductsMapper {
```
Comment (EMazzaglia) [inline src/products/mappers/products.mapper.ts]: You can use `class-transformer` to do this. I suggest you read the documentation for it :) 
https://github.com/typestack/class-transformer

File: src/products/services/products.service.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+import { GetAllProductsException } from '@products/exceptions/get-all-products.exception';
+import { ProductsMapper } from '@products/mappers/products.mapper';
+
+@Injectable()
+export class ProductsService {
+  private readonly logger: Logger = new Logger(ProductsService.name);
+
+  constructor(
+    @InjectRepository(ProductEntity)
+    private repository: Repository<ProductEntity>,
```
Comment (EMazzaglia) [inline src/products/services/products.service.ts]: `repository` as a variable name is not clear enough (we are going to have 25 repositories probably. WDYT about naming it as `productRepository`?

File: src/products/controllers/products.controller.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import { Controller, Get } from '@nestjs/common';
+import { ProductDTO } from '@products/dtos/product.dto';
+import { ProductsService } from '@products/services/products.service';
+
+@Controller('products')
+export class ProductsController {
+  constructor(private service: ProductsService) {}
+
+  @Get()
+  async getAll(): Promise<ProductDTO[]> {
```
Comment (EMazzaglia) [inline src/products/controllers/products.controller.ts]: You dont need to use the keyword `async` here.

Comment (EMazzaglia) [inline src/products/controllers/products.controller.ts]: You should't retrieve a list of productDTO. Instead, you should receive a DTO and return a list of Products (your entities).
DTO stands for Data Transfer Object, transport the data from the outworld to our app. Then we need to map that to our entities if and only if the object is valid. That's why `class-validator` or any other way of validating objects comes to play at that point.

File: src/products/tests/controllers/products.controller.spec.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    it('should throw a GetAllProductsException if service fails', async () => {
+      jest.spyOn(productService, 'getAll').mockImplementationOnce(() => {
+        throw new GetAllProductsException('');
+      });
+
+      await expect(productsController.getAll()).rejects.toThrowError(
+        GetAllProductsException,
+      );
+
+      expect(productService.getAll).toHaveBeenCalledWith();
```
Comment (EMazzaglia) [inline src/products/tests/controllers/products.controller.spec.ts]: What's this line for?

File: src/products/tests/fakes/products.fake.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    name.title(),
+    random.words(10),
+    datatype.number(1000000),
+    datatype.number(100),
+    internet.url(),
+  );
+};
+
+export const createFakeProductEntityArray = (size = 5): ProductEntity[] => {
+  const products: ProductEntity[] = [];
```
Comment (EMazzaglia) [inline src/products/tests/fakes/products.fake.ts]: You are failing into a DRY violation repeating logic here and on this `createFakeProductDTOArray` function. 
Extract the logic and reuse it in both places.

File: src/products/tests/services/products.service.spec.ts [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+      response.forEach((product) => {
+        expect(product.name).toBeDefined();
+        expect(product.description).toBeDefined();
+        expect(product.price).toBeDefined();
+        expect(product.stock).toBeDefined();
+        expect(product.image).toBeDefined();
+      });
+
+      expect(productsRepository.find).toHaveBeenCalledWith();
```
Comment (EMazzaglia) [inline src/products/tests/services/products.service.spec.ts]: You should use, `toHaveBeenCalled` or `toHaveBeenCalledTimes(1)`

File: src/products/tests/services/products.service.spec.ts [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    it('should throw a GetAllProductsException if repository fails', async () => {
+      jest.spyOn(productsRepository, 'find').mockImplementationOnce(() => {
+        throw new InternalServerErrorException();
+      });
+
+      await expect(productsService.getAll()).rejects.toThrowError(
+        GetAllProductsException,
+      );
+
+      expect(productsRepository.find).toHaveBeenCalledWith();
```
Comment (EMazzaglia) [inline src/products/tests/services/products.service.spec.ts]: Same comment

File: src/products/tests/services/products.service.spec.ts [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+
+    it('should return empty array if no products were found', async () => {
+      jest.spyOn(productsRepository, 'find').mockResolvedValueOnce([]);
+
+      const response = await productsService.getAll();
+
+      expect(response).toBeDefined();
+      expect(response.length).toBe(0);
+
+      expect(productsRepository.find).toHaveBeenCalledWith();
```
Comment (EMazzaglia) [inline src/products/tests/services/products.service.spec.ts]: same comment

**Review by EMazzaglia — CHANGES_REQUESTED** [review]
Good overall work, I left some comments. 

I would like to see e2e tests as well and how do you manage those.

File: src/products/entities/product.entity.ts (hunk 6 above)
Comment (Darvand) [reply src/products/entities/product.entity.ts]: Is assigning the same type as the `ProductEntity.name` has

File: src/products/mappers/products.mapper.ts (hunk 8 above)
Comment (Darvand) [reply src/products/mappers/products.mapper.ts]: That's new for me. Great!

File: src/products/tests/controllers/products.controller.spec.ts (hunk 11 above)
Comment (Darvand) [reply src/products/tests/controllers/products.controller.spec.ts]: It checks if method `getAll` from `productService` was called without arguments

File: src/products/exceptions/get-all-products.exception.ts (hunk 7 above)
Comment (Darvand) [reply src/products/exceptions/get-all-products.exception.ts:1]: Why do you say camelCase in some filenames? I'm naming every file with hypen-case

File: src/products/controllers/products.controller.ts (hunk 10 above)
Comment (Darvand) [reply src/products/controllers/products.controller.ts]: Without the `async` word, the test related to the throw error exception will fail

File: src/database/migrations/1655309590353-create-product-table.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class createProductTable1655309590353 implements MigrationInterface {
+  name = 'createProductTable1655309590353';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `CREATE TABLE "products" ("id" uuid NOT NULL DEFAULT uuid_generate_v4(), "created_at" TIMESTAMP NOT NULL DEFAULT now(), "updated_at" TIMESTAMP NOT NULL DEFAULT now(), "name" character varying NOT NULL, "description" text, "price" integer NOT NULL, "stock" integer NOT NULL DEFAULT '0', "image" character varying, CONSTRAINT "UQ_4c9fb58de893725258746385e16" UNIQUE ("name"), CONSTRAINT "PK_0806c755e0aca124e67c0cf6d7d" PRIMARY KEY ("id"))`,
```
Comment (slasia) [inline src/database/migrations/1655309590353-create-product-table.ts:8]: I think you should set the max length of the strings fields. What do you think?

File: src/products/services/products.service.ts [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+import { ProductEntity } from '@products/entities/product.entity';
+import { GetAllProductsException } from '@products/exceptions/get-all-products.exception';
+
+@Injectable()
+export class ProductsService {
+  private readonly logger: Logger = new Logger(ProductsService.name);
+
+  constructor(
+    @InjectRepository(ProductEntity)
+    private productRepository: Repository<ProductEntity>,
```
Comment (slasia) [inline src/products/services/products.service.ts]: I would recommend you creating a repository layer in order to separate the database logic from the bussiness.

Example

```node
@EntityRepository(Product)
export class ProductRepository extends Repository<Product> {
// your custom db operations here
}
```
And in your service

```node
 @InjectRepository(CategoryRepository) private categoriesRepo: CategoryRepository
```

So now you have your database logic in your "category.repository.ts" file and the service only calls it without dealing with ORM/SQL functions.

**Review by slasia — DISMISSED** [review]

File: src/products/entities/product.entity.ts (hunk 6 above)
Comment (EMazzaglia) [reply src/products/entities/product.entity.ts]: So we could make use of the productEntity as an interface to create a product right? This comment is related with a well known code smell 
"Long parameter list". That code smell is present when you have a large list of parameters to construct an object or call a method. 
To solve this you should receive an object that has inside all the params needed to instantiate the product.

File: src/products/exceptions/get-all-products.exception.ts (hunk 7 above)
Comment (EMazzaglia) [reply src/products/exceptions/get-all-products.exception.ts:1]: You are right, I got confused! my bad

File: src/products/dtos/product.dto.ts (hunk 1 above)
Comment (Darvand) [reply src/products/dtos/product.dto.ts]: Yes but I remove the DTO for now because I'm not currently requesting information. Just returning it using entities. On future, when create a DTO, I will use `class-validator` for sure

File: src/products/entities/product.entity.ts (hunk 2 above)
Comment (Darvand) [reply src/products/entities/product.entity.ts:7]: Is not working with PostgreSQL. It was throwing me errors for that attribute:
image

File: src/products/services/products.service.ts (hunk 17 above)
Comment (Darvand) [reply src/products/services/products.service.ts]: Well, yes, I can create the repository layer but I think I can use the injection right now because I'm not using something complex or special query. I can simple use the default methods from `Repository<ProductEntity>`
Or what do you think? Is not a problem to create the repository also

Comment (slasia) [reply src/products/services/products.service.ts]: Yes, It will be useful thinking that the app can scale later

File: src/database/migrations/1655309590353-create-product-table.ts (hunk 16 above)
Comment (Darvand) [reply src/database/migrations/1655309590353-create-product-table.ts:8]: I think is a bit complex just for something that we can try to guard from the backend service

File: package.json (hunk 4 above)
Comment (Darvand) [reply package.json]: Done!

File: src/products/dtos/product.dto.ts (hunk 5 above)
Comment (Darvand) [reply src/products/dtos/product.dto.ts]: I remove the DTOs on this PR because I'm not longer using it. In later PRs, I will use `class-validator` for sure

File: src/products/controllers/products.controller.ts (hunk 10 above)
Comment (Darvand) [reply src/products/controllers/products.controller.ts]: I remove the DTOs on this PR because I'm not longer using it. In later PRs, I will use class-validator for sure

File: src/products/tests/fakes/products.fake.ts (hunk 12 above)
Comment (Darvand) [reply src/products/tests/fakes/products.fake.ts]: I remove the other method that create ProductDTO; so, this method is now the only

File: src/products/tests/services/products.service.spec.ts (hunk 13 above)
Comment (Darvand) [reply src/products/tests/services/products.service.spec.ts]: I use the `toHaveBeenCalledWith` to evaluate two things: that has been called with no arguments and that I call it. But I can change it with `toHaveBeenCalledTimes(1)` with no problem

File: src/products/tests/services/products.service.spec.ts (hunk 14 above)
Comment (Darvand) [reply src/products/tests/services/products.service.spec.ts]: I use the toHaveBeenCalledWith to evaluate two things: that has been called with no arguments and that I call it. But I can change it with toHaveBeenCalledTimes(1) with no problem

File: src/products/tests/services/products.service.spec.ts (hunk 15 above)
Comment (Darvand) [reply src/products/tests/services/products.service.spec.ts]: I use the toHaveBeenCalledWith to evaluate two things: that has been called with no arguments and that I call it. But I can change it with toHaveBeenCalledTimes(1) with no problem

File: src/products/mappers/products.mapper.ts (hunk 3 above)
Comment (Darvand) [reply src/products/mappers/products.mapper.ts]: Yes but I remove the DTO for now because I'm not currently requesting information. Just returning it using entities. But I will have this in mind for the next time

File: src/products/services/products.service.ts (hunk 17 above)
Comment (Darvand) [reply src/products/services/products.service.ts]: Done!

File: src/products/dtos/product.dto.ts (hunk 5 above)
Comment (EMazzaglia) [reply src/products/dtos/product.dto.ts]: Nice, thanks

File: src/products/entities/product.entity.ts (hunk 6 above)
Comment (EMazzaglia) [reply src/products/entities/product.entity.ts]: @Darvand ?

File: src/products/services/products.service.ts [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+@Injectable()
+export class ProductsService {
+  private readonly logger: Logger = new Logger(ProductsService.name);
+
+  constructor(
+    @InjectRepository(ProductsRepository)
+    private readonly productRepository: ProductsRepository,
+  ) {}
+
+  async getAll(): Promise<ProductEntity[]> {
```
Comment (EMazzaglia) [inline src/products/services/products.service.ts]: I would use just Product as the name of the class.

File: src/products/tests/controllers/products.controller.spec.ts [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+
+      response.forEach((product) => {
+        expect(product.name).toBeDefined();
+        expect(product.description).toBeDefined();
+        expect(product.price).toBeDefined();
+        expect(product.stock).toBeDefined();
+        expect(product.image).toBeDefined();
+      });
+
+      expect(productService.getAll).toHaveBeenCalledWith();
```
Comment (EMazzaglia) [inline src/products/tests/controllers/products.controller.spec.ts]: Hello there, As I mentioned earlier, please update the `toHaveBeenCalledWith` to `toHaveBeenCalled`

File: src/products/tests/controllers/products.controller.spec.ts [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+
+    it('should return empty array if no product were found', async () => {
+      jest.spyOn(productService, 'getAll').mockResolvedValueOnce([]);
+
+      const response = await productsController.getAll();
+
+      expect(response).toBeDefined();
+      expect(response.length).toBe(0);
+
+      expect(productService.getAll).toHaveBeenCalledWith();
```
Comment (EMazzaglia) [inline src/products/tests/controllers/products.controller.spec.ts]: Hello there, As I mentioned earlier, please update the toHaveBeenCalledWith to toHaveBeenCalled

File: src/products/tests/controllers/products.controller.spec.ts (hunk 11 above)
Comment (EMazzaglia) [reply src/products/tests/controllers/products.controller.spec.ts]: Hello there, As I mentioned earlier, please update the toHaveBeenCalledWith to toHaveBeenCalled

File: src/products/tests/fakes/products.fake.ts (hunk 12 above)
Comment (EMazzaglia) [reply src/products/tests/fakes/products.fake.ts]: LGTM

File: src/products/tests/products.e2e-spec.ts [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+    app = module.createNestApplication();
+    repository = module.get<Repository<ProductEntity>>(
+      getRepositoryToken(ProductEntity),
+    );
+    await app.init();
+  });
+
+  describe('GET /products', () => {
+    it('should return empty array because there are no products', async () => {
+      const response = await request(app.getHttpServer()).get('/products');
```
Comment (EMazzaglia) [inline src/products/tests/products.e2e-spec.ts]: If you watch this line, you see that you are repeating the 

```
app.getHttpServer()
```
 in all your tests. It's better to just declare a global variable and make use of that in all the tests to aovid falling into a DRY violation.

File: src/products/tests/products.e2e-spec.ts [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+      expect(response.statusCode).toEqual(HttpStatus.OK);
+      expect(response.body).toEqual([]);
+    });
+
+    it('should return five products', async () => {
+      const expectedProducts = createFakeProductEntityArray();
+      await repository.insert(expectedProducts);
+      const response = await request(app.getHttpServer()).get('/products');
```
Comment (EMazzaglia) [inline src/products/tests/products.e2e-spec.ts]: Same comment, this line introduces a DRY violation.

**Review by slasia — DISMISSED** [review]

File: src/products/tests/services/products.service.spec.ts [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+    it('should throw a GetAllProductsException if repository fails', async () => {
+      jest.spyOn(productsRepository, 'findAll').mockImplementationOnce(() => {
+        throw new InternalServerErrorException();
+      });
+
+      await expect(productsService.getAll()).rejects.toThrowError(
+        GetAllProductsException,
+      );
+
+      expect(productsRepository.findAll).toHaveBeenCalledWith();
```
Comment (EMazzaglia) [inline src/products/tests/services/products.service.spec.ts]: Replace this with toHaveBeenCalled since this function is called with no parameters.

**Review by EMazzaglia — CHANGES_REQUESTED** [review]
Hello good rework, I left more comments please address them and reach me when you are done :D Ty

File: src/products/entities/product.entity.ts (hunk 6 above)
Comment (Darvand) [reply src/products/entities/product.entity.ts]: But how this will solve the "Long parameter list"? If I have an object as param that contains all the attributes, then I have to initiate this object somewhere with all params also. So, I don't really understand how this will solve that

File: src/products/services/products.service.ts (hunk 18 above)
Comment (Darvand) [reply src/products/services/products.service.ts]: Well, yes, I can make this change

File: src/products/tests/controllers/products.controller.spec.ts (hunk 19 above)
Comment (Darvand) [reply src/products/tests/controllers/products.controller.spec.ts]: Done!

File: src/products/tests/controllers/products.controller.spec.ts (hunk 20 above)
Comment (Darvand) [reply src/products/tests/controllers/products.controller.spec.ts]: Done!

File: src/products/tests/products.e2e-spec.ts (hunk 21 above)
Comment (Darvand) [reply src/products/tests/products.e2e-spec.ts]: Done!

File: src/products/tests/products.e2e-spec.ts (hunk 22 above)
Comment (Darvand) [reply src/products/tests/products.e2e-spec.ts]: Done!

File: src/products/tests/services/products.service.spec.ts (hunk 23 above)
Comment (Darvand) [reply src/products/tests/services/products.service.spec.ts]: Done!

File: src/products/tests/products.e2e-spec.ts (hunk 21 above)
Comment (Darvand) [reply src/products/tests/products.e2e-spec.ts]: Done!

File: src/products/tests/products.e2e-spec.ts (hunk 22 above)
Comment (Darvand) [reply src/products/tests/products.e2e-spec.ts]: Done!

File: src/products/tests/services/products.service.spec.ts (hunk 23 above)
Comment (Darvand) [reply src/products/tests/services/products.service.spec.ts]: Done!

File: src/products/services/products.service.ts (hunk 18 above)
Comment (Darvand) [reply src/products/services/products.service.ts]: Done!

File: src/products/entities/product.entity.ts (hunk 6 above)
Comment (EMazzaglia) [reply src/products/entities/product.entity.ts]: Because we are encapsulating all the parameters in an object which will serve the purpose of an interface for the method. That will give us the advantage to reuse that interface whenever we need to call the function.
You can read more about Long Parameter List code smell [here ](https://refactoring.guru/smells/long-parameter-list)

File: src/products/tests/services/products.service.spec.ts (hunk 13 above)
Comment (EMazzaglia) [reply src/products/tests/services/products.service.spec.ts]: This method has no params so its more accurate to go with `toHaveBeenCalled`

File: src/products/tests/services/products.service.spec.ts (hunk 15 above)
Comment (EMazzaglia) [reply src/products/tests/services/products.service.spec.ts]: Answered this in previous comment :) Good job

File: src/products/services/products.service.ts (hunk 18 above)
Comment (EMazzaglia) [reply src/products/services/products.service.ts]: good job!

**Review by EMazzaglia — DISMISSED** [review]
LGTM

**Review by EMazzaglia — APPROVED** [review]

**Review by slasia — APPROVED** [review]
